core/src/network_manager/manager.rs

```diff
@@ -427,6 +491,35 @@ impl NetworkManagerExecutor {
         }
     }
 
+    // ------------------------------
+    // | Identify Protocol Handlers |
+    // ------------------------------
+
+    /// Handle a message from the Identify protocol
+    async fn handle_identify_event(
+        &mut self,
+        event: IdentifyEvent,
+    ) -> Result<(), NetworkManagerError> {
+        // Update the local peer's public IP address if it has not already been discovered
+        if let IdentifyEvent::Received { info, .. } = event {
+            if !self.discovered_identity {
+                // Replace the port if the discovered NAT port is incorrect
+                let mut local_addr = info.observed_addr;
+                replace_port(&mut local_addr, self.p2p_port);
```

Comment:
might be my lack of networking knowledge here... but if the observed IP from the Identify protocol says we communicated on port `x`, shouldn't we stick with that instead of changing it to `self.p2p_port` and gossiping that around to peers?

I guess my real question here is why would these 2 ports (observed vs `self.p2p_port`) be in mismatch? Could it be that our outbound messages come from a different port than the one we're receiving on?

Reply:
Yeah good question; one thing is that [NAT](https://en.wikipedia.org/wiki/Network_address_translation) boxes in between the connection endpoints may swizzle these ports to their own internal mapping. This is especially the case in e.g. a symmetric NAT. 

The observed address may then be the address that the NAT has assigned to the connection, not the address that the local node has bound to. Anecdotally, I noticed this happening during testing